Tabnine/src/main/java/com/tabnine/lifecycle/pushtosignin/PushToSignIn.kt

```diff
@@ -1,59 +1,49 @@
 package com.tabnine.lifecycle.pushtosignin
-import com.intellij.openapi.application.ApplicationManager
-import com.intellij.openapi.components.ServiceManager
-import com.tabnineCommon.binary.requests.config.StateResponse
-import com.tabnineCommon.capabilities.CapabilitiesService
-import com.tabnineCommon.capabilities.Capability
-import com.tabnineCommon.capabilities.CapabilityNotifier
-import com.tabnineCommon.lifecycle.BinaryStateChangeNotifier
-import com.tabnineCommon.lifecycle.BinaryStateService
-import com.tabnineCommon.lifecycle.PluginInstalled
+
+import com.intellij.notification.Notification
+import com.tabnineCommon.lifecycle.BinaryStateSingleton
 import java.util.concurrent.atomic.AtomicBoolean
 
 class PushToSignIn {
     private var started = AtomicBoolean(false)
+    private lateinit var innerState: PushToSignInState
+    private var notification: Notification? = null
+
     fun start() {
         if (started.getAndSet(true)) {
             return
         }
-        CapabilityNotifier.subscribe(
-            CapabilityNotifier { state ->
-                if (state.isEnabled(Capability.FORCE_REGISTRATION)) {
-                    transition()
-                }
-            }
-        )
-        val connection = ApplicationManager.getApplication()?.messageBus?.connect()
-        connection?.subscribe(
-            BinaryStateChangeNotifier.STATE_CHANGED_TOPIC,
-            BinaryStateChangeNotifier { response ->
-                transition(response)
-            }
-        )
 
-        PluginInstalled.subscribe(
-            PluginInstalled { installed ->
-                if (installed) {
-                    transition()
-                }
-            }
-        )
+        innerState = PushToSignInState()
+        innerState.useState {
+            transition(it.isForceRegistration, it.isNewInstallation, it.isLoggedIn)
+        }
     }
-    private fun transition(state: StateResponse? = null) {
-        val loggedIn = if (state != null) state.isLoggedIn else ServiceManager.getService(BinaryStateService::class.java).lastStateResponse?.isLoggedIn
 
-        val isNewInstallation = PluginInstalled.isNewInstallation
+    @Synchronized
+    private fun transition(
+        forceRegistration: Boolean,
+        isNewInstallation: Boolean?,
+        loggedIn: Boolean?
+    ) {
         when {
-            !CapabilitiesService.getInstance().isCapabilityEnabled(Capability.FORCE_REGISTRATION) -> return
+            !forceRegistration -> return
```

Comment:
is this still relevant?

Reply:
Yes, the force registration flag is still alive